File: packages/js/src/sources/console/recorder.ts
import type { LogStore } from '../../core/store.js';
import type { LogEntryInit, LogKind, LogLevel, LogPart } from '../../core/types.js';
import { formatArguments } from './format.js';
import { DEFAULT_CAPTURE_OPTIONS, snapshotValue, type CaptureOptions } from './snapshot.js';
import { formatTable } from './table.js';

/** The console methods lognal records. */
export type ConsoleMethod =
	| 'log'
	| 'info'
	| 'warn'
	| 'error'
	| 'debug'
	| 'trace'
	| 'dir'
	| 'dirxml'
	| 'table'
	| 'group'
	| 'groupCollapsed'
	| 'groupEnd'
	| 'count'
	| 'countReset'
	| 'time'
	| 'timeLog'
	| 'timeEnd'
	| 'assert'
	| 'clear';

export const CONSOLE_METHODS: readonly ConsoleMethod[] = [
	'log',
	'info',
	'warn',
	'error',
	'debug',
	'trace',
	'dir',
	'dirxml',
	'table',
	'group',
	'groupCollapsed',
	'groupEnd',
	'count',
	'countReset',
	'time',
	'timeLog',
	'timeEnd',
	'assert',
	'clear'
];

export interface RecorderOptions extends CaptureOptions {
	/** Whether `console.clear` removes the entries from the store. */
	clearStore: boolean;
}

export const DEFAULT_RECORDER_OPTIONS: RecorderOptions = {
	...DEFAULT_CAPTURE_OPTIONS,
	clearStore: true
};

const DEFAULT_LABEL = 'default';

const now = (): number => {
	return typeof performance !== 'undefined' ? performance.now() : Date.now();
};

const labelOf = (value: unknown): string => {
	return value === undefined ? DEFAULT_LABEL : String(value);
};

const formatDuration = (milliseconds: number): string => {
	return `${Number(milliseconds.toFixed(3))} ms`;
};

const text = (value: string, extra: Partial<LogPart> = {}): LogPart => {
	return { type: 'text', text: value, ...extra } as LogPart;
};

/**
 * Turns console calls into store entries.
 *
 * A recorder keeps the state the Console Standard gives a console: the count map, the timer
 * table and the group stack. Everything is captured synchronously, when the method is called.
 */
export class ConsoleRecorder {
	private options: RecorderOptions;
	private readonly counts = new Map<string, number>();
	private readonly timers = new Map<string, number>();
	private groups: number[] = [];

	constructor(
		private readonly store: LogStore,
		options: Partial<RecorderOptions> = {}
	) {
		this.options = { ...DEFAULT_RECORDER_OPTIONS, ...options };
	}

	setOptions(options: Partial<RecorderOptions>): void {
		this.options = { ...this.options, ...options };
	}

	/**
	 * Records one call. `stack` is the caller's stack trace, which only `trace` uses.
	 */
	record(method: ConsoleMethod, args: readonly unknown[], stack?: string): void {
		switch (method) {
			case 'log':
			case 'info':
			case 'warn':
			case 'error':
			case 'debug':
				this.add(method, formatArguments(args, this.capture));
				break;
			case 'dir':
				this.add('log', [{ type: 'value', value: this.capture(args[0]) }]);
				break;
			case 'dirxml':
				this.add(
					'log',
					args.length > 0
						? args.flatMap((value, index): LogPart[] => [
								...(index > 0 ? [text(' ')] : []),
								{ type: 'value', value: this.capture(value) }
							])
						: [text('undefined', { token: 'null' })]
				);
				break;
			case 'trace':
				this.trace(args, stack);
				break;
			case 'table':
				this.table(args);
				break;
			case 'group':
			case 'groupCollapsed':
				this.group(args, method === 'groupCollapsed');
				break;
			case 'groupEnd':
				this.groups = this.groups.slice(0, -1);
				break;
			case 'count':
				this.count(args[0]);
				break;
			case 'countReset':
				this.countReset(args[0]);
				break;
			case 'time':
				this.time(args[0]);
				break;
			case 'timeLog':
				this.timeLog(args[0], args.slice(1));
				break;
			case 'timeEnd':
				this.timeEnd(args[0]);
				break;
			case 'assert':
				this.assert(args);
				break;
			case 'clear':
				this.clear();
				break;
			default:
				break;
		}
	}

	private readonly capture = (value: unknown) => {
		return snapshotValue(value, this.options);
	};

	private add(
		level: LogLevel,
		parts: LogPart[],
		kind: LogKind = 'message',
		init: Partial<LogEntryInit> = {}
	): void {
		this.store.append({ level, kind, parts, groups: [...this.groups], ...init });
	}

	private trace(args: readonly unknown[], stack?: string): void {
		const parts: LogPart[] =
			args.length > 0 ? formatArguments(args, this.capture) : [text('console.trace')];

		if (stack) {
			const frames = stack
				.split('\n')
				.map((frame) => frame.trim())
				.filter(Boolean)
				.map((frame) => `    ${frame}`)
				.join('\n');

			parts.push(text(`\n${frames}`, { token: 'muted' }));
		}

		this.add('log', parts);
	}

	private table(args: readonly unknown[]): void {
		const [data, properties] = args;

		if (data === null || typeof data !== 'object') {
			this.add('log', formatArguments(args, this.capture));

			return;
		}

		const node = snapshotValue(data, { ...this.options, maxDepth: 2 });
		const columns = Array.isArray(properties) ? properties.map((key) => String(key)) : undefined;
		const table = formatTable(node, columns);

		// A table keeps its rows whole; a table wider than the viewer scrolls sideways.
		this.add(
			'log',
			table === null ? [{ type: 'value', value: node }] : [text(table, { wrap: false })]
		);
	}

	private group(args: readonly unknown[], collapsed: boolean): void {
		const parts =
			args.length > 0
				? formatArguments(args, this.capture).map((part) =>
						part.type === 'text' ? { ...part, style: { ...part.style, bold: true } } : part
					)
				: [text(collapsed ? 'console.groupCollapsed' : 'console.group', { style: { bold: true } })];
		const [entry] = this.store.append({
			level: 'log',
			kind: 'group',
			parts,
			groups: [...this.groups],
			collapsed
		});

		if (entry) {
			this.groups = [...this.groups, entry.id];
		}
	}

	private count(label: unknown): void {
		const key = labelOf(label);
		const count = (this.counts.get(key) ?? 0) + 1;

		this.counts.set(key, count);
		this.add('info', [text(`${key}: ${count}`)]);
	}

	private countReset(label: unknown): void {
		const key = labelOf(label);

		if (this.counts.has(key)) {
			this.counts.set(key, 0);
		} else {
			this.add('warn', [text(`Count for '${key}' does not exist`)]);
		}
	}

	private time(label: unknown): void {
		const key = labelOf(label);

		if (this.timers.has(key)) {
			this.add('warn', [text(`Timer '${key}' already exists`)]);

			return;
		}

		this.timers.set(key, now());
	}

	private timeLog(label: unknown, data: readonly unknown[]): void {
		const key = labelOf(label);
		const start = this.timers.get(key);

		if (start === undefined) {
			this.add('warn', [text(`Timer '${key}' does not exist`)]);

			return;
		}

		const parts: LogPart[] = [text(`${key}: ${formatDuration(now() - start)}`)];

		for (const value of data) {
			parts.push(text(' '));
			parts.push(
				typeof value === 'string' ? text(value) : { type: 'value', value: this.capture(value) }
			);
		}

		this.add('log', parts);
	}

	private timeEnd(label: unknown): void {
		const key = labelOf(label);
		const start = this.timers.get(key);

		if (start === undefined) {
			this.add('warn', [text(`Timer '${key}' does not exist`)]);

			return;
		}

		this.timers.delete(key);
		this.add('info', [text(`${key}: ${formatDuration(now() - start)}`)]);
	}

	private assert(args: readonly unknown[]): void {
		const [condition, ...data] = args;

		if (condition) {
			return;
		}

		if (data.length === 0) {
			this.add('error', [text('Assertion failed')]);

			return;
		}

		if (typeof data[0] === 'string') {
			data[0] = `Assertion failed: ${data[0]}`;
		} else {
			data.unshift('Assertion failed');
		}

		this.add('error', formatArguments(data, this.capture));
	}

	private clear(): void {
		this.groups = [];

		if (this.options.clearStore) {
			this.store.clear();
		}

		this.add('log', [text('Console was cleared', { token: 'muted' })], 'system');
	}
}

File: packages/flutter/lib/src/sources/console/recorder.dart
import 'package:lognal/src/core/store.dart';
import 'package:lognal/src/core/types.dart';
import 'package:lognal/src/sources/console/capture.dart';
import 'package:lognal/src/sources/console/format.dart';
import 'package:lognal/src/sources/console/table.dart';

/// The logging calls lognal records.
enum ConsoleMethod {
  /// An ordinary message.
  log,

  /// A message worth noticing.
  info,

  /// Something that may be a problem.
  warn,

  /// Something that is a problem.
  error,

  /// Detail a developer asked for.
  debug,

  /// A message with the stack trace of its caller under it.
  trace,

  /// One value, opened rather than described.
  dir,

  /// A collection drawn as a table.
  table,

  /// The start of a group.
  group,

  /// The start of a group that begins closed.
  groupCollapsed,

  /// The end of the innermost open group.
  groupEnd,

  /// How many times this label has been counted.
  count,

  /// Sets a label's count back to zero.
  countReset,

  /// Starts a timer.
  time,

  /// Reports how long a timer has been running, without stopping it.
  timeLog,

  /// Reports how long a timer ran and stops it.
  timeEnd,

  /// An error message, unless the condition holds.
  assertCondition,

  /// Removes every entry.
  clear,
}

/// Every method a recorder answers to.
const List<ConsoleMethod> consoleMethods = ConsoleMethod.values;

/// What a recorder captures and what `clear` does.
class RecorderOptions {
  /// Creates the options.
  const RecorderOptions({this.capture = defaultCaptureOptions, this.clearStore = true});

  /// The limits a captured value is taken under.
  final CaptureOptions capture;

  /// Whether [ConsoleMethod.clear] removes the entries from the store.
  final bool clearStore;

  /// A copy with the fields given here replaced.
  RecorderOptions copyWith({CaptureOptions? capture, bool? clearStore}) {
    return RecorderOptions(
      capture: capture ?? this.capture,
      clearStore: clearStore ?? this.clearStore,
    );
  }
}

/// The options a recorder starts with.
const RecorderOptions defaultRecorderOptions = RecorderOptions();

const String _defaultLabel = 'default';

final Stopwatch _clock = Stopwatch()..start();

double _now() => _clock.elapsedMicroseconds / 1000;

String _labelOf(Object? value) => value == null ? _defaultLabel : value.toString();

String _formatDuration(double milliseconds) {
  final String text = milliseconds.toStringAsFixed(3);

  return '${text.contains('.') ? text.replaceAll(RegExp(r'\.?0+$'), '') : text} ms';
}

TextPart _text(String value, {StyleToken? token, LogTextStyle? style, bool wrap = true}) {
  return TextPart(value, token: token, style: style, wrap: wrap);
}

/// Turns logging calls into store entries.
///
/// A recorder keeps the state a console has: the count map, the timer table and
/// the group stack. Everything is captured synchronously, when the method is
/// called.
class ConsoleRecorder {
  /// Creates a recorder that writes into a store.
  ConsoleRecorder(this.store, {this.options = defaultRecorderOptions});

  /// The store the entries go into.
  final LogStore store;

  /// The options in use. Replacing them applies to the next call.
  RecorderOptions options;

  final Map<String, int> _counts = <String, int>{};
  final Map<String, double> _timers = <String, double>{};
  List<int> _groups = <int>[];

  /// Records one call. [stack] is the caller's stack trace, which only
  /// [ConsoleMethod.trace] and an error use.
  void record(ConsoleMethod method, List<Object?> args, {StackTrace? stack}) {
    switch (method) {
      case ConsoleMethod.log:
        _add(LogLevel.log, formatArguments(args, _capture));
      case ConsoleMethod.info:
        _add(LogLevel.info, formatArguments(args, _capture));
      case ConsoleMethod.warn:
        _add(LogLevel.warn, formatArguments(args, _capture));
      case ConsoleMethod.debug:
        _add(LogLevel.debug, formatArguments(args, _capture));
      case ConsoleMethod.error:
        _error(args, stack);
      case ConsoleMethod.dir:
        _add(LogLevel.log, <LogPart>[ValuePart(_capture(args.isEmpty ? null : args.first))]);
      case ConsoleMethod.trace:
        _trace(args, stack);
      case ConsoleMethod.table:
        _table(args);
      case ConsoleMethod.group:
        _group(args, false);
      case ConsoleMethod.groupCollapsed:
        _group(args, true);
      case ConsoleMethod.groupEnd:
        _groups = _groups.isEmpty ? _groups : _groups.sublist(0, _groups.length - 1);
      case ConsoleMethod.count:
        _count(args.isEmpty ? null : args.first);
      case ConsoleMethod.countReset:
        _countReset(args.isEmpty ? null : args.first);
      case ConsoleMethod.time:
        _time(args.isEmpty ? null : args.first);
      case ConsoleMethod.timeLog:
        _timeLog(args.isEmpty ? null : args.first, args.skip(1).toList());
      case ConsoleMethod.timeEnd:
        _timeEnd(args.isEmpty ? null : args.first);
      case ConsoleMethod.assertCondition:
        _assert(args);
      case ConsoleMethod.clear:
        _clear();
    }
  }

  ValueNode _capture(Object? value) => captureValue(value, options.capture);

  void _add(
    LogLevel level,
    List<LogPart> parts, {
    LogKind kind = LogKind.message,
    bool collapsed = false,
  }) {
    store.add(
      LogEntryInit(
        parts: parts,
        level: level,
        kind: kind,
        groups: List<int>.of(_groups),
        collapsed: collapsed,
      ),
    );
  }

  /// An error call, with the stack trace the caller passed or the one the error
  /// carries.
  void _error(List<Object?> args, StackTrace? stack) {
    if (stack == null || args.isEmpty) {
      _add(LogLevel.error, formatArguments(args, _capture));

      return;
    }

    // `error(exception, stackTrace)` is how Dart reports a failure, so the two
    // become one entry whose value carries the trace.
    final ValueNode captured = _capture(args.first);

    if (captured.kind != ValueKind.error) {
      _add(LogLevel.error, <LogPart>[
        ...formatArguments(args, _capture),
        _text('\n${_framesOf(stack)}', token: StyleToken.muted),
      ]);

      return;
    }

    final List<LogPart> rest = formatArguments(args.skip(1).toList(), _capture);
    final List<LogPart> parts = <LogPart>[
      ValuePart(captured.copyWith(stack: captured.stack ?? _framesOf(stack))),
      if (rest.isNotEmpty) _text(' '),
      ...rest,
    ];

    _add(LogLevel.error, parts);
  }

  void _trace(List<Object?> args, StackTrace? stack) {
    final List<LogPart> parts = args.isEmpty
        ? <LogPart>[_text('trace')]
        : formatArguments(args, _capture);
    final String? frames = stack == null ? null : _framesOf(stack);

    if (frames != null && frames.isNotEmpty) {
      parts.add(_text('\n$frames', token: StyleToken.muted));
    }

    _add(LogLevel.log, parts);
  }

  void _table(List<Object?> args) {
    final Object? data = args.isEmpty ? null : args.first;
    final Object? properties = args.length > 1 ? args[1] : null;

    if (data == null || data is String || data is num || data is bool) {
      _add(LogLevel.log, formatArguments(args, _capture));

      return;
    }

    final ValueNode node = captureValue(data, options.capture.copyWith(maxDepth: 2));
    final List<String>? columns = properties is List<Object?>
        ? properties.map((Object? key) => '$key').toList()
        : null;
    final String? table = formatTable(node, columns);

    // A table keeps its rows whole; a table wider than the viewer scrolls
    // sideways.
    _add(
      LogLevel.log,
      table == null ? <LogPart>[ValuePart(node)] : <LogPart>[_text(table, wrap: false)],
    );
  }

  void _group(List<Object?> args, bool collapsed) {
    final List<LogPart> parts = args.isEmpty
        ? <LogPart>[_text(collapsed ? 'group' : 'group', style: const LogTextStyle(bold: true))]
        : formatArguments(args, _capture)
              .map(
                (LogPart part) => part is TextPart
                    ? part.copyWith(
                        style: (part.style ?? const LogTextStyle()).copyWith(bold: true),
                      )
                    : part,
              )
              .toList();
    final LogEntry? entry = store.add(
      LogEntryInit(
        parts: parts,
        kind: LogKind.group,
        groups: List<int>.of(_groups),
        collapsed: collapsed,
      ),
    );

    if (entry != null) {
      _groups = <int>[..._groups, entry.id];
    }
  }

  void _count(Object? label) {
    final String key = _labelOf(label);
    final int count = (_counts[key] ?? 0) + 1;

    _counts[key] = count;
    _add(LogLevel.info, <LogPart>[_text('$key: $count')]);
  }

  void _countReset(Object? label) {
    final String key = _labelOf(label);

    if (_counts.containsKey(key)) {
      _counts[key] = 0;
    } else {
      _add(LogLevel.warn, <LogPart>[_text("Count for '$key' does not exist")]);
    }
  }

  void _time(Object? label) {
    final String key = _labelOf(label);

    if (_timers.containsKey(key)) {
      _add(LogLevel.warn, <LogPart>[_text("Timer '$key' already exists")]);

      return;
    }

    _timers[key] = _now();
  }

  void _timeLog(Object? label, List<Object?> data) {
    final String key = _labelOf(label);
    final double? start = _timers[key];

    if (start == null) {
      _add(LogLevel.warn, <LogPart>[_text("Timer '$key' does not exist")]);

      return;
    }

    final List<LogPart> parts = <LogPart>[_text('$key: ${_formatDuration(_now() - start)}')];

    for (final Object? value in data) {
      parts
        ..add(_text(' '))
        ..add(value is String ? _text(value) : ValuePart(_capture(value)));
    }

    _add(LogLevel.log, parts);
  }

  void _timeEnd(Object? label) {
    final String key = _labelOf(label);
    final double? start = _timers.remove(key);

    if (start == null) {
      _add(LogLevel.warn, <LogPart>[_text("Timer '$key' does not exist")]);

      return;
    }

    _add(LogLevel.info, <LogPart>[_text('$key: ${_formatDuration(_now() - start)}')]);
  }

  void _assert(List<Object?> args) {
    final Object? condition = args.isEmpty ? null : args.first;

    if (condition == true) {
      return;
    }

    final List<Object?> data = args.skip(1).toList();

    if (data.isEmpty) {
      _add(LogLevel.error, <LogPart>[_text('Assertion failed')]);

      return;
    }

    if (data.first is String) {
      data[0] = 'Assertion failed: ${data.first}';
    } else {
      data.insert(0, 'Assertion failed');
    }

    _add(LogLevel.error, formatArguments(data, _capture));
  }

  void _clear() {
    _groups = <int>[];

    if (options.clearStore) {
      store.clear();
    }

    _add(LogLevel.log, <LogPart>[
      _text('Log was cleared', token: StyleToken.muted),
    ], kind: LogKind.system);
  }
}

/// A stack trace indented the way a stack under a message is drawn, with the
/// frames of lognal itself left out.
String _framesOf(StackTrace stack) {
  final List<String> frames = stack
      .toString()
      .split('\n')
      .map((String frame) => frame.trim())
      .where((String frame) => frame.isNotEmpty && !frame.contains('package:lognal/'))
      .map((String frame) => '    $frame')
      .toList();

  return frames.join('\n');
}
